include/pybind11/stl.h

```diff
@@ -242,14 +263,14 @@ template<typename T> struct optional_caster {
         if (!src) {
             return false;
         } else if (src.is_none()) {
-            value = {};  // nullopt
+            optional_reset<T>::reset(value);
             return true;
         }
         value_conv inner_caster;
         if (!inner_caster.load(src, convert))
             return false;
 
-        value.emplace(cast_op<typename T::value_type>(inner_caster));
+        value.emplace(std::move(cast_op<typename T::value_type>(inner_caster)));
```

Comment:
Not safe: type casters can return a lvalue reference to a variable they don't own.  We might, however, be able to coax type casters cast ops in rvalue context into giving up an rvalue reference.